nomos-core/chain-defs/src/block/mod.rs

```diff
@@ -46,12 +74,114 @@ impl Proposal {
+        signing_key: &ed25519_dalek::SigningKey,
+    ) -> Result<Self, Error>
+    where
+        Tx: Transaction<Hash = TxHash>,
+    {
+        if transactions.len() > MAX_TRANSACTIONS {
+            return Err(Error::Signing(format!(
+                "Too many transactions (max {MAX_TRANSACTIONS})"
+            )));
+        }
+
+        let block_root =
+            Self::calculate_content_id_from_parts(&transactions, service_reward.as_ref());
+
+        let header = Header::new(parent_block, block_root, slot, proof_of_leadership);
+
+        let signature = header.sign(signing_key)?;
+
+        Ok(Self {
             header,
+            signature,
+            service_reward,
             transactions,
+        })
+    }
+
+    pub fn recover(
+        header: Header,
+        transactions: Vec<Tx>,
+        service_reward: Option<Tx>,
+        signature: ed25519_dalek::Signature,
+    ) -> Result<Self, Error>
+    where
+        Tx: Transaction<Hash = TxHash>,
+    {
+        if !header.is_valid_bedrock_version() {
+            return Err(Error::Signing("Invalid header version".to_owned()));
+        }
+
+        if transactions.len() > MAX_TRANSACTIONS {
+            return Err(Error::Signing(format!(
+                "Too many transactions (max {MAX_TRANSACTIONS})"
+            )));
+        }
+
+        let calculated_content_id =
+            Self::calculate_content_id_from_parts(&transactions, service_reward.as_ref());
+        if header.block_root() != &calculated_content_id {
+            return Err(Error::Signing(
+                "Block root mismatch with transactions".to_owned(),
+            ));
         }
+
+        let leader_public_key = header.leader_proof().leader_key();
+        let header_bytes = <Header as SerdeOp>::serialize(&header)?;
+
+        leader_public_key
+            .verify(&header_bytes, &signature)
+            .map_err(|e| Error::Signing(format!("Invalid signature: {e}")))?;
+
+        Ok(Self {
+            header,
+            signature,
+            service_reward,
+            transactions,
+        })
+    }
+
+    fn calculate_content_id_from_parts(
+        transactions: &[Tx],
+        service_reward: Option<&Tx>,
+    ) -> ContentId
+    where
+        Tx: Transaction<Hash = TxHash>,
+    {
+        let mut all_transactions = Vec::new();
+        if let Some(reward) = service_reward {
+            all_transactions.push(reward);
+        }
+        all_transactions.extend(transactions.iter());
+
+        if all_transactions.is_empty() {
+            return ContentId::from([0; 32]);
+        }
+
+        let merkle_leaves: Vec<merkle::MerkleNode<TxHash>> = all_transactions
+            .iter()
+            .enumerate()
+            .map(|(i, tx)| {
+                if i % 2 == 0 {
+                    merkle::MerkleNode::Left(tx.hash())
+                } else {
+                    merkle::MerkleNode::Right(tx.hash())
+                }
+            })
+            .collect();
```

Comment:
mmm no? Not sure whether we need to pad the tree (added a comment on the spec), but you can take the  necessary building blocks from https://github.com/logos-co/nomos-pocs/blob/master/emmarin/cl/cl/src/ds/merkle.rs

Reply:
> spec also says blake2

FYI, that's because I asked in a comment there after this thread :)